auth_mellon_util.c

```diff
@@ -1835,7 +1957,130 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
     return url;
 }
 
+/* Join array of strings with separator
+ *
+ * Parameters:
+ *  pool      memory allocation pool
+ *  strings   NULL terminated array of NULL terminated strings
+ *  separator string to insert between elements
+ *
+ * Example:
+ *  char *strings[] = {"one", "two", NULL};
+ *  am_str_join(pool, strings, ", ")
+ *
+ *  would return "one, two"
+ */
+
+char *am_str_join(apr_pool_t *pool, char **strings, const char *separator)
+{
+    char *str = NULL;
+    char *dst = NULL;
+    char **p;
+    apr_size_t i;
+    apr_size_t len;
+    apr_size_t alloc_len;
+    apr_size_t n_strings = 0;
+    apr_size_t total_string_len = 0;
+    apr_size_t separator_len = strlen(separator);
+
+    /*
+     * Iterate over string array, compute total number of characters
+     * and a count of how many strings
+     */
+    for (total_string_len = 0, n_strings = 0, p = strings; *p; p++) {
+        total_string_len += strlen(*p);
+        n_strings++;
+    }
+
+    /*
+     * The allocation length is the length of all the strings
+     * plus the length of all the separators plus 1 for null terminator
+     */
+    alloc_len = total_string_len +
+        (n_strings > 0 ? separator_len*(n_strings-1) : 0) + 1;
+
+    str = apr_palloc(pool, alloc_len);
+    for (i = 0, dst = str; i < n_strings; i++) {
+        /* If this is not the first string insert a separator */
+        if (i > 0) {
+            memcpy(dst, separator, separator_len);
+            dst += separator_len;
+        }
+        /* Copy the string */
+        len = strlen(strings[i]);
+        memcpy(dst, strings[i], len);
+        dst += len;
+
+    }
+    *dst++ = 0;                 /* NULL terminate */
+
+    return str;
+}
+
 #ifdef HAVE_ECP
+
+/* String representation of ECPServiceOptions bitmask
+ *
+ * ECPServiceOptions is a bitmask of flags. Return a comma separated string
+ * of all the flags. If any bit in the bitmask is unaccounted for an
+ * extra string will be appended of the form "(unknown bits = x)".
+ *
+ * Parameters:
+ *  pool    memory allocation pool
+ *  options bitmask of PAOS options
+ */
+char *am_ecp_service_options_str(apr_pool_t *pool, ECPServiceOptions options)
+{
+    int i = 0;
+    char *names[10];            /* Warning, increase this if flags are added,
```

Comment:
Doesn't this need to be `const char *` to avoid some warnings, since you are assigning constant strings to it?